[1단계 - 로또 구현] 몽이(황재웅) 미션 제출합니다.

wodnd0131 opened the pull request:
## 프로젝트 개요

- 페어 프로그래밍 교대 시간은 약 20분입니다.
- 공동 목표는 테스트에 기반하여 검증을 진행하는 것입니다 (네오의 강의 방식대로 진행했습니다)
  - 멍구(페어)는 객체 설계에 중점을 두었습니다
  - 몽이(글쓴이)는 직접적인 기존 코드 참고를 최소화했습니다

## 주요 기술적 의사결정

### 클래스 설계
- Record Class를 활용하여 Lotto 객체를 구현했습니다
- Wallet(지갑) 객체를 도입하여 Lotto 일급 컬렉션을 관리했습니다

### 테스트 전략
- getter 대신 validate를 통해 객체가 직접 검증하도록 구성했습니다
- 각 도메인 객체 별로 단위 테스트를 구현했습니다
- 타 도메인 테스트 코드에서 검증된 경우는 추가로 검증하지 않았습니다
(예: 로또에서 검증된 것을 지갑에서 다시 검증하지 않도록 했습니다)

## 고민했던 부분
- DependencyConfig 사용
의존성을 일관적으로 관리하기 위해 도입했으나, 서비스 크기가 작아 아직 필요성이 느껴지지 않아 삭제하기로 했습니다.
 
- View 계층의 인터페이스 도입 고려
기존 과제에서는 외부 라이브러리를 사용했기에 외부와 접하는 인터페이스를 활용했었습니다.
그러나 본 과제에서는 자바 기본 라이브러리이기에 적용하지 않았습니다.

- Scanner vs BufferedReader 선택
서비스 규모가 작고 입력받는 값이 적다고 판단하여 Scanner를 채택했습니다.

- DTO는 필요한가?
네오의 강의를 들으며 고민하게 된 부분입니다.
기존의 도메인으로 분리되어 있던 MatchCount와 Profit은 도메인 정보와 관련되었기에 DTO가 아닌 도메인으로 재배치했습니다.

- 불규칙한 로또번호는 어떻게 테스트 할 것인가?
처음에는 로또 번호에 대해 사이즈로만 검증을 하고, 내부 값에 대한 검증이 부족했습니다.
네오의 강의를 보고 Random 함수를 통해 시드값을 지정하는 방식으로 수정하였습니다.
추후에 같은 문제가 있다면 인터페이스를 활용해볼까합니다.

- RandomWrapper는 도메인을 포함해도 괜찮은가?
기존의 RandomWrapper에서는 클래스 내부에 직접적인 도메인 정보를 포함했습니다
(예를 들어, 로또 사이즈 등).
하지만, 해당 클래스는 도메인과 무관한 utill 클래스이며, Random 라이브러리를 포장하기 위한 Wrapper이기에, 도메인에 관한 정보를 분리하여 매개변수로 받도록 했습니다.

## 학습 포인트
1. "객체지향의 오해와 사실"을 읽으며 느낀, 행동이 우선이라는 규칙을 설계 단계에서 적용할 것입니다
2. 단위 테스트를 적극 활용했습니다 (네오의 강의 방식을 그대로 따라 해보았습니다)
3. 페어 프로그래밍을 통해 지난 과제와의 차이점을 이해했습니다
---
[Prolog 링크](https://prolog.techcourse.co.kr/studylogs/3841)

Commit:
docs: 기능 요구사항 정리

Commit:
feat: Application 시작점 선언

Commit:
feat: Lotto 객체 선언

Commit:
feat: RandomsWrapper로 랜덤 번호 생성

Commit:
feat: Amount 선언

- 로또 구입 개수를 계산한다.

Commit:
feat: Wallet을 선언

- 고객의 구매한 로또를 관리한다.

Commit:
feat: Controller 선언

Commit:
feat: MatchCountDto를 선언

- 당첨 개수와 보너스 여부를 저장

Commit:
feat: MatchStatistics를 선언

- 당첨 결과에 맞게 변환

Commit:
style: 코딩 컨벤션 적용

Commit:
refactor: 메소드 명 개선 및 코드 구조 정리

- 일부 메소드 명 정정
- 조건문을 메소드로 분리
- Common 패키지 추가

Commit:
feat: Calculator를 선언

- matchCount를 토대로 통계 결과를 반환한다.
- 수익률을 계산한다.

Commit:
feat: Profit Dto를 선언

Commit:
feat: View를 선언

- 입출력 담당 클래스를 선언한다.

Commit:
feat: Parser를 추가한다.

- 입력값을 숫자로 변환한다.
- 입력값을 숫자 List로 반환한다.

Commit:
feat: Wallet의 로또를 전달하기 위해 가공

-toString을 Override하여 수정한다.
-MatchStatics에서 Key에 맞는 String 값을 반환하도록 메소드를 구현한다.

Commit:
feat: Controller.run()을 구현한다.

- DependencyConfig를 통해 의존성 관리

Commit:
refactor: Lotto에서 Match Count를 계산하도록 수정

- Wallet에서 진행하던 일부를 Lotto에게 시켜서 반환하도록 변경했다.
- Constant를 통해 매직 넘버 관리

Commit:
refactor: Calculator를 LottoService로 변경

- 역할이 도메인보다 서비스에 적합하여 변경한다.
- 내부의 메소드명을 정정한다.

Commit:
test: LottoTest에 단위테스트를 추가

- 로또 정렬과 MatchCount()에 대한 테스트를 추가한다.
- InputParser에 String 조건을 상수로 변경한다.

Commit:
feat: 보너스 번호 검증 추가

- 매직넘버, 상수 추가한다.
- 불필요한 테스트 삭제

Commit:
feat: Error 발생 시 다시 질문하기 구현

- MatchStatistics Test를 추가한다.
- Error Message에 Prefix를 추가한다.

Commit:
chore: 스타일 컨밴션 적용

sure-why-not commented on README.md:
```diff
@@ -1,3 +1,76 @@
+# java-lotto 로또 미션 저장소
 
-로또 미션 저장소
+##  기능 요구사항
```
todo list와 실행 결과 예시까지! 꼼꼼하게 작성해주셨네요 👍🏻

wodnd0131 replied:
라라! 만나게 되어 반갑습니다~😄
이번 미션 리뷰해주셔서 감사해요!
부족한 실력이지만 많이 배울 수  있도록 노력하겠습니다!🫡

sure-why-not commented on src/main/java/lotto/Application.java:
```diff
@@ -0,0 +1,13 @@
+        Controller controller = dependencyConfig.getController();
+        controller.run();
+    }
+}
```
<img width="43" alt="스크린샷 2025-02-14 오전 12 44 13" src="https://github.com/user-attachments/assets/9beb6684-ddff-4802-9fef-040cfff090af" />

파일 마지막 라인에 개행을 자동으로 추가해주는 기능을 사용해봐도 좋을 것 같아요!
https://jojoldu.tistory.com/673

sure-why-not commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,77 @@
+
+    private void validateSize(List<Integer> numbers) {
+        if (isCorrectedSize(numbers)) {
+            throw new IllegalArgumentException(ERROR_LOTTO_SIZE);
```
`ERROR_LOTTO_SIZE` 라는 값이 숫자일 줄 알았는데 `[ERROR] 로또 번호는 6개여야 합니다` 라는 문구를 담고 있네요.
ErrorMessage 의 값들을 사용하는 쪽에서 더 명확하게 인지할 수 있는 네이밍으로 수정해보는건 어떨까요? 😀

wodnd0131 replied:
Constant에서 LOTTO_SIZE라는 상수에서는 int로 관리되기 때문에 헷갈릴 수 있는 문제라 느꼈습니다.

Prefix 최적화와 함께 위 문제를 해결하기 위해 ErrorMessage를 enum 클래스로 전환했습니다.(+ 패키지도 Constant로 이동)

또한, 에러 메세지 네이밍을 좀더 직관적이도록 수정했습니다.
ERROR_MONEY_ZERO -> ERROR_MONEY_LESS_THEN_STANDARD

이를 통해 ERROR_INCORRECT_LOTTO_SIZE.getMessage() 의 형태로 호출되기에, 반환 타입을 좀 더 명시적으로 표시했습니다.

sure-why-not commented on src/main/java/lotto/domain/MatchStatistics.java:
```diff
@@ -0,0 +1,49 @@
+            return MATCH_BONUS;
+        }
+
+        for (MatchStatistics statistics : values()) {
+            if (statistics.number == matchNumber) {
+                return statistics;
+            }
+        }
```
규칙1이 지켜지지 않았어요 🥲
```
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.를 지키며 구현한다.
- 예를 들어 while문 안에 if문이 있으면 들여쓰기는 2이다.
```

wodnd0131 replied:
페어 프로그래밍을 하며, 기능 완성만 생각하여 놓친  유지 보수에 문제입니다.
미처 검토하지 않은 몇몇 코드들을 숨길 수 없네요,,,
 
다음 페어 프로그래밍에서는 개발 절차를 좀 더 체계화해야 겠습니다.
지금은  목표 -> 구현 -> 테스트 ->(전체 완성)-> 유지 보수의 단계였지만,
 테스트 이후 바로바로 유지 보수를 진행하고, 
 다른 기능 구현에 조급해하지 않도록 주의하겠습니다.🥲

sure-why-not commented on src/test/java/lotto/domain/WalletTest.java:
```diff
@@ -0,0 +1,18 @@
+
+class WalletTest {
+
+    @DisplayName("toString 가공 테스트")
+    @Test
+    void testToString() {
+        Amount amount = new Amount(5000);
+
+        Wallet wallet = new Wallet(amount);
+
+        System.out.println(wallet);
+    }
```
요건 어떤 목적의 테스트 인가요?
항상 예상하는 동일한 결과를 얻는지 테스트할 수 있나요~?

wodnd0131 replied:
해당 코드 또한, 다시 작성해야 함을 검토하지 않았던 부분이네요.

본래의 목적은, Wallet에서 저장된 List<Lotto>에 대해 toString으로 출력할 경우,
Lotto의 목록 나열을 요구 사항과 동일하게 한 줄 씩 리스트 배열을 출력하는지 확인하기 위함이었습니다.

이때 Record를 통해 Lotto를 표현했기에 이에 대한 학습 테스트의 목적이 좀 더 컸던 것 같습니다. 때문에 중요도가 낮았다 느꼈는지 잊어버리는 실수를 범했네요🫠

목적에 맞게 Test 코드 정정했습니다.

sure-why-not commented on src/test/java/lotto/utill/RandomsWrapperTest.java:
```diff
@@ -0,0 +1,20 @@
+class RandomsWrapperTest {
+
+    @Test
+    @DisplayName(" ")
+    void test() {
```
엇 요친구는 더 구체적인 테스트명이 필요해요 🥲 (깜빡하신거죠?!)

wodnd0131 replied:
해당 테스트 클래스는 제거하였습니다..!
Wrapper 클래스의 테스트는 학습 테스트만 요구되어야 한다고 생각합니다.
때문에, 클래스를 다시 살펴보니 도메인에 관한 요소가 섞여서 클래스의 기능이 애매해졌다고 느꼈습니다.

이 때문에, Wrapper 클래스 목적에 맞게 Random 함수 호출 외의 요소를 Lotto로 배제하였고, 이에 따라 별도의 Test가 요구되지 않는다고 느껴 제거하였습니다.

반대로, Lotto의 경우 기능이 추가적으로 배정되었기에 이에 따라 Test 코드 개선을 하였습니다.

sure-why-not commented on src/test/java/lotto/common/utill/InputParserTest.java:
```diff
@@ -0,0 +1,42 @@
+        assertThat(parseToInt(str)).isEqualTo(1000);
+    }
+
+    @DisplayName("문자를 포함할 때, 예외가 발생한다.")
```
정상케이스와 예외케이스 모두 테스트를 잘 추가해주셨네요 👍🏻

DisplayName에 어떤 기능에 대한 테스트인지 적어주시면 다른 사람이 어떤 테스트인지 파악하기 쉬울 것 같아요! 
아래는 예시일 뿐이에요 ㅎㅎ
```suggestion
    @DisplayName("String 입력값을 Int로 변환할 때, 문자를 포함할 경우, 예외가 발생한다.")
```

그리고 parseToList 에 대한 테스트도 추가해보는건 어떨까요? 😀

wodnd0131 replied:
다시 보니 "문자를 포함할 때, 예외가 발생한다."라는 이름은 정확히 어떤 상황인지 
한번에 이해하기 어렵네요. 문자라는 범위는 너무 넓어서 오해가 생길 것 같아요.
좀 더 문제 상황을 특정하기 용이한 이름으로 지칭할 필요성을 배우게 되네요..

parseToList에 대한 테스트도 추가했습니다!

sure-why-not commented on src/main/java/lotto/common/constant/Constant.java:
```diff
@@ -0,0 +1,7 @@
+package lotto.common.constant;
+
+public class Constant {
```
1. Constant와 ErrorMessage를 따로 클래스로 분리한 이유가 있을까요?
2. 로또에 대한 규칙 상수들을 모아둔걸까요? 조금더 구체적인 이름을 지어보는건 어떨까요? ㅎㅎ

wodnd0131 replied:
Constant와 ErrorMessage를 따로 클래스로 분리한 이유는 ErrorMessage의 경우 Enum 클래스로 관리 하고자 했기 때문이었습니다. 다만, 시간 관리 실패로 미숙한 모습을 보이네요,,

Constant와 달리 ErrorMessage의 경우 다음 두 이유로 분리 하였습니다.
1. Prefix를 통해 에러임을 명시한다.
2. 단순히 상수로써 관리한다는 목적 외에도 문제 상황을 확인할 수 있어야 한다.

이 두 부분은 Enum 클래스로 변경하며 반영하였습니다!

sure-why-not commented on src/main/java/lotto/domain/Amount.java:
```diff
@@ -0,0 +1,43 @@
+public class Amount {
+    private static final int LOTTO_PRIZE = 1000;
+
+    private final int money;
```
원시값 포장 굳👍🏻
그런데 amount(금액)이 money(돈)을 갖고 있는게 조금 어색한 것 같아요🤔

money(돈)이 amount(금액)을 갖고있거나
amount(금액)이 value(값)을 갖고 있는건 어떤가요?

wodnd0131 replied:
단순히 getAmount()라는 메소드의 필요성을 느끼며 객체를 만들며 
세부적인 고민이 부족했네요..

Amont는 금액/돈을 입력 받아 만들어집니다.
이때 주된 역할은 다음과 같습니다.

1. 입력된 돈이 로또 구매에 적합한지 검증
2. 구매할 수 있는 로또의 개수를 반환

이때, 사용자의 금액으로 "몇 개의 로또를 구매할 수 있는가?" 라는 것이 해당 객체의 주된 책임이라 생각합니다.
때문에 이에 맞게 로또 매장에서 결제를 담당하는 Cashier라 다시 명명하였습니다.

sure-why-not commented on src/main/java/lotto/common/utill/RandomsWrapper.java:
```diff
@@ -0,0 +1,20 @@
+    public static List<Integer> getRandomNumbers() {
+        List<Integer> list = new ArrayList<>();
+
+        while (list.size() < 6) {
+            int randomInt = (int)(Math.random() * 45) + 1;
+            if (!list.contains(randomInt)) {
+                list.add(randomInt);
+            }
+        }
```
로또 번호의 정책과 동일한 값들이라면 여기서도 Constant 의 값들을 사용해보는건 어떨까요?

wodnd0131 replied:
수정하며 해당 기능은 도메인과 관련되었다 판단하여,
해당 메서드 getRandomNumbers()를 호출할 때 Constant를 활용해 인자로 넘겨주도록 수정했습니다!

sure-why-not reviewed (changes requested):
몽이 안녕하세요! 리뷰어 라라입니다 😁
첫 미션인데도 전체적으로 잘 구현해주셨네요 👍🏻
이번 미션에서 필요한 부분에 집중하여 코멘트 남겼으니 확인 부탁드려요!
화이팅입니다~ 💪🏻

Commit:
refactor: 누락된 Application의 마지막 라인 개행을 추가

Commit:
refactor: ErrorMessage를 enum으로 변경

- 패키지 위치를 exception -> constant로 이동한다.
- ErrorMessage을 직관적이도록 수정한다.
- .getMessage() 호출을 통해, 에러 메세지를 반환함을 명시적으로 표현한다.

Commit:
refactor: indent가 2인 문제 수정

- matchNumber에 대한 매직넘버를 상수화한다.

Commit:
refactor: RandomsWrapper의 랜덤 추첨을 Random를 활용하도록 조정

- Random 라이브러리는 시드값이 존재하여 테스트에 용이하다.
- inputParser의 for문을 스트림으로 변경한다.

Commit:
refactor: Lotto의 생성 방식을 조정

- RandomsWrapper.getRandomIntList()로부터 도메인 요소를 분리한다.
- Lotto.generateLotto()를 통해 랜덤 번호를 생성한다.
- Lotto 생성자는 검증 이전에 중복 제거와 정렬을 진행한다.

Commit:
refactor: Lotto와 Wallet Test를 개선

-불필요해진 RandomsWrapperTest를 삭제한다.
-LottoTest에서 상수를 통해 도메인 기준을 따르도록 한다.
-Wallet의 모호했던 testToString를 개선한다.

Commit:
refactor: 의미가 모호한 Dto를 Domain으로 전환

-Profit과 MatchCount는 요구사항에 관한 것으로 Domain으로 전환한다.

Commit:
refactor: 의미가 모호한 Test 메소드 명 정정

- ErrorMessage와 Amount의 검증 메서드 명도 정정한다.

Commit:
refactor: 불필요한 DependencyConfig 제거

-현재 서비스는 크기가 작아 불필요하다 판단했다.

Commit:
refactor: ErrorMessage에 에러 타입 추가

Commit:
refactor: Amount의 클래스명을 Cashier로 변경

Commit:
feat: Lotto 클래스에 bouns 번호에 대한 검증 추가

Commit:
fix: 로또 생성시 중복제거로 발생하는 갯수 문제 해결

- 번호 생성시 중복제거 처리를 추가한다.

Commit:
refactor: paresToList에 대한 예외처리 추가

-ParseToListTest 또한 추가한다.

sure-why-not commented on src/main/java/lotto/controller/Controller.java:
```diff
@@ -0,0 +1,100 @@
+
+    }
+
+    private int requestInt(String output) {
```
숫자를 입력받는 것이니, `requestNumber` 는 어떤가요? ㅎㅎ

wodnd0131 replied:
수정했습니다. int는 너무 개발자를 위한 표현이군요.
좀 더 도메인 친화적인 네이밍이 좋다는 걸 배웠네요!

Commit:
test: RandomWrapper 테스트를 추가

-Random에 관한 학습 테스트 추가한다.
-RandomsWrapper를 RandomWrapper로 개명한다.

sure-why-not commented on src/main/java/lotto/controller/Controller.java:
```diff
@@ -0,0 +1,100 @@
+        return lottoService.convertToMap(matchCount);
+    }
+
+    private Cashier requestAmount() {
+        while (true) {
+            try {
+                int money = requestInt("구입금액을 입력해 주세요.");
+                return new Cashier(money);
+            } catch (IllegalArgumentException e) {
+                outputView.print(e.getMessage());
+            }
+        }
+
+    }
```
금액을 요청하는 메서드인데 캐셔가 반환되고있네요 🤔

wodnd0131 replied:
흠,, 이전 PR 에서 Amount 클래스 자체에 대해 고민했지만, 이와 관련된 부수적인 부분들은 생각하지 못했네요.

```java
     Cashier cashier = requestAmount();
        Wallet wallet = new Wallet(cashier);
```
이 부분도 Wallet을 캐셔로 만든다 혹 캐셔가 Wallet에 포함된다는 것이 어색한 것 같습니다.

초기 설계 구조 혹 객체의 명칭을 바꿀 때는 이러한 맥락 또한 고려해야 하는군요.
각 객체의 의미를 고려해서 다음과 같이 수정했습니다.
```java
        Cashier cashier = requestToCashier();
        int numberOfLotto = cashier.getNumberOfLotto();
        Wallet wallet = new Wallet(numberOfLotto);
```
코드가 조금 길어졌지만, 각 객체 이전보다 직관적인 것 같습니다.🙂

Commit:
refactor: MatchStatistics를 MatchInfo로 개명

Commit:
refactor: Constant를 BusniessRule로 개명

sure-why-not commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,93 @@
+
+    public MatchCount matchCount(Lotto matchLotto, int bonus) {
+        int count = 0;
+        for (int i : matchLotto.numbers()) {
+            if (isContainsBonus(i)) {
+                count++;
+            }
+        }
```
다음 단계에선 이부분도 규칙을 지켜보아요!
```
규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.를 지키며 구현한다.
- 예를 들어 while문 안에 if문이 있으면 들여쓰기는 2이다.
```

sure-why-not commented on src/test/java/lotto/domain/CashierTest.java:
```diff
@@ -0,0 +1,35 @@
+
+class CashierTest {
+    @Test
+    @DisplayName(" ")
```
👀

sure-why-not commented on src/test/java/lotto/common/utill/InputParserTest.java:
```diff
@@ -0,0 +1,89 @@
+
+class InputParserTest {
+
+    @Nested
+    class ParseToIntTest {
+        @DisplayName("String 입력값을 Int 값으로 변환한다.")
```
테스트명을 구체적으로 작성하니 어떤 테스트인지 훨씬 파악하기 쉬워졌네요!
`@Nested`로 묶으니 어떤 기능에 대한 테스트들인지 구분하기도 쉬워졌네요  👍🏻

Nested에도 DisplayName를 달아서 어떤 기능 테스트인지 작성할 수 있어요 ㅎㅎ

```suggestion
    @Nested
    @DisplayName("String 입력값을 Int 값으로 변환할 때")
    class ParseToIntTest {
        @DisplayName("정상적으로 변환한다.")
```

sure-why-not commented on src/test/java/lotto/service/LottoServiceTest.java:
```diff
@@ -0,0 +1,29 @@
+import lotto.domain.MatchInfo;
+import lotto.domain.Profit;
+
+class LottoServiceTest {
```
테스트를 어떤 기준으로 추가하시나요?
LottoService.isProfit() 테스트는 없네요 🤔

wodnd0131 replied:
테스트 작성에 명확한 기준이 없었던 것 같습니다.
코딩 중 실행 결과가 필요하다고 느껴졌을 때 만들었던 것 같네요.
도중에, 몇몇 함수에서 "이정도는 검증하지 않아도 괜찮다"라는 마음이 전반적인 퀄리티를 낮춰버린 것 같습니다.

테스트 코드 작성에서
1차적으로 도메인에서 중요한 검증을,
2차적으로 객체의 메소드 별 테스트 코드 작성함에 좀 더 의식해봐야겠습니다.😾

sure-why-not commented on src/test/java/lotto/domain/MatchInfoTest.java:
```diff
@@ -0,0 +1,18 @@
+class MatchInfoTest {
+
+    @Test
+    @DisplayName("getMatchStatistics 정상 작동 테스트")
```
요 테스트도 없네요 👀
- MatchStatistics.getMatchData()
- MatchStatistics.isMatchBonus()

sure-why-not commented on src/test/java/lotto/domain/WalletTest.java:
```diff
@@ -0,0 +1,22 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class WalletTest {
```
Wallet의 핵심 기능인 matchCount() 테스트도 없네요 🥲

wodnd0131 replied:
중요한 테스트가 누락되었네요..! 🤤
사용자의 로또와 지난주 당첨 로또를 비교하는 것이 matchCount()의 역할입니다.
이때 기존의 코드에서는 Wallet과 Lotto 둘 다 이 메소드를 가지고 있습니다.

Lotto.matchCount()의 경우, 실제 로또의 비교를 진행합니다.
반면, Wallet.matchCount()의 경우, Wallet 내부의 lottolist의 각 로또에게 Lotto.matchCount()의 결과값을 요청하고, 이것들을 그룹화하여 반환하는 책임을 가집니다.

이 때문에, 먼저 Wallet.matchCount()의 경우, 역할의 맞지 않는 이름이라 판단하여 matchCount->getMatchCountList로 수정했습니다. 그리고 로또 리스트의 모든 값에 대해 비교하는 지 확인하는 테스트를 추가했습니다.

그리고, 실제 핵심 기능인 Lotto.matchCount()에 대한 테스트를 LottoTest에 별도로 추가했습니다..!

sure-why-not reviewed (approved):
몽이, 피드백 반영을 꼼꼼하게 해주셨네요! 💯
테스트가 부족한 부분 코멘트 남겨두었는데, 다음 미션에서 반영해보아요!
미션마다 핵심 키워드에 집중하며 우테코 기간동안 많은 것을 얻어가셨으면 좋겠어요 😀
첫번째 미션 고생 많으셨고, 다음에 또 만나요~ 👋🏻
